src/packet/key/public.rs

```diff
@@ -0,0 +1,506 @@
+use byteorder::{BigEndian, ByteOrder, WriteBytesExt};
+use md5::Md5;
+use sha1_checked::{Digest, Sha1};
+
+use crate::{
+    crypto::{self, hash::HashAlgorithm, public_key::PublicKeyAlgorithm},
+    errors::Result,
+    packet::{Signature, SignatureConfigBuilder, SignatureType, Subpacket, SubpacketData},
+    types::{KeyId, KeyVersion, Mpi, PublicKeyTrait, PublicParams, SecretKeyTrait, Tag, Version},
+};
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct PublicKey(PubKeyInner);
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct PublicSubkey(PubKeyInner);
+
+impl PublicKey {
+    /// Create a new `PublicKey` packet from underlying parameters.
+    pub fn new(
+        packet_version: Version,
+        version: KeyVersion,
+        algorithm: PublicKeyAlgorithm,
+        created_at: chrono::DateTime<chrono::Utc>,
+        expiration: Option<u16>,
+        public_params: PublicParams,
+    ) -> Result<Self> {
+        let inner = PubKeyInner::new(
+            packet_version,
+            version,
+            algorithm,
+            created_at,
+            expiration,
+            public_params,
+        )?;
+        Ok(Self(inner))
+    }
+
+    /// Parses a `PublicKeyKey` packet from the given slice.
+    pub fn from_slice(packet_version: Version, input: &[u8]) -> Result<Self> {
+        let inner = PubKeyInner::from_slice(packet_version, input)?;
+        Ok(Self(inner))
+    }
+
+    pub fn sign<F>(&self, key: &impl SecretKeyTrait, key_pw: F) -> Result<Signature>
```

Comment:
I wonder if this function (and its subkey sibling, below) should both be named `bind`, to be slightly more self-explanatory.

Reply:
Moving this point to https://github.com/rpgp/rpgp/issues/367, it is separate from this PR